week1/homework/src/Contact.js

```diff
@@ -0,0 +1,36 @@
+export default class Contact {
+  constructor(name, phone) {
+    this.name = name;
+    this.phone = phone;
+  }
+
+  get phone() {
+    return this._phone;
+  }
+
+  set phone(phone) {
+    if (
```

Comment:
hi @jawkha, it's quite hard to read this bit of code. Here's some tips:

- you can turn this `if` around and save yourself a lot of code. if you condition under which it is a valid phone number, then you could test against that. any other situation (`else`), it's invalid.
- you can also do the exact opposite, which I would call 'exit early':
```js
if (someCondition) {
  return console.log("Please input a valid telephone number");
}
if (someOtherCondition) {
  return console.log("Please input a valid telephone number");
}
this._phone = phone;
return;
```
This keeps the indentation low, easier to read.
- instead of having long conditions, put them first in a variable that says what you are testing:
```js
const isNumber = typeof phone !== 'number';
const hasRightLength = String(phone).length === 8;

if (isNumber && hasRightLength) {
  this._phone = phone;
}
```
- you could parse the phone number into a number first with `const phoneNumber = Number(phone);` and from there on work with `phoneNumber` instead

- what value does this line return `return (this._phone = phone);`? Hard to say right, that is also why it's [actively discouraged](https://eslint.org/docs/rules/no-return-assign)

Reply:
Hi @godspeedelbow, thank you for your comments. I have rewritten the phone number validation portion of the file following your advice. To ensure that any non-string and non-number types are not accepted, I set a condition for accepting the input. Then I convert it into number. I also define variables that store the various conditions for validating the input.
   ```
    set phone(phone) {
    // changes the type of input to Number (first tried to apply type conversion without any conditions but then it was failing a test where a number within an array, e.g. [32143211] was being input as a phone number)
    if (typeof phone === "string" || typeof phone === "number") {
      const phoneNumber = Number(phone);

      // to check that the input after conversion is a valid positive integer
      const isValidNumber = typeof phoneNumber === "number" && Number.isInteger(phoneNumber) && phoneNumber > 0;
      
      // to check that the number has 8 digits
      const hasRightLength = String(phoneNumber).length === 8;
      
      if (isValidNumber && hasRightLength) {
        this._phone = phoneNumber;
        return this._phone;
      }
    } else {
      console.log("Please input a valid phone number.");
    }
  }
   ```
Here is a problem though. The way I have currently set it up, the console.log message is only displayed if the input is a non-string or non-number. One way to ensure that the message would show up for all invalid inputs is to use `var` instead of `const` and then defining the rest of the function outside that "if" block from where I can access this `var`. The second way to do it is to duplicate the console.log message as an "else" for the inner "if" block. Somehow, both of those don't look like good choices to me. Is there another way to handle this?
Thank you again for all your help.

Reply:
- you can turn the statement by inverting the conditional, this gives you one indent less and makes it easier to read, what I called 'exit early':

```js
if (typeof phone !== "string" && typeof phone !== "number") {
  console.log("Please input a valid phone number.");
}

const phoneNumber = Number(phone);
...
```

- btw if you add `js` after the three backticks when opening a code block it will parse the block as javascript, gives it code highlighting :)

- sometimes you have situations where you need to return the same thing on different lines. it's okay, quite common actually. you can just define a `const invalidMessage = 'please provide ...';` at the top of the function and `return console.log(invalidMessage)` instead. Another option is rethinking whether this function should log anything. Maybe it should return `true` or `false`, indicating if it was successful. In case, of `false` the callee that tried to set the phone number can log an error message.

- if somebody provides an array, you have two options: rejecting it or try to parse it. If you want to parse it, you could create a loose separate `function parsePhoneToNumber(number) { ... }` that returns a `Number`. Then you can `const number = parsePhoneToNumber(phone);` at the beginning. From there you know it's a number.

cheers!

Reply:
Thank you so much for taking the time to explain things. Your feedback has been extremely helpful. I think it's a lot easier to read compared to the prior versions.
```js
set phone(phone) {
    const invalidPhoneMessage = "Please input a valid phone number.";
    // only accepts inputs with type of string or number
    if (typeof phone !== "string" || typeof phone !== "number") {
      console.log(invalidPhoneMessage);
      return;
    } 
      // changes the type of input to Number
      const phoneNumber = Number(phone);

      // to check that the input after conversion is a valid positive integer
      const isValidNumber =
        typeof phoneNumber === "number" &&
        Number.isInteger(phoneNumber) &&
        phoneNumber > 0;

      // to check that the number has 8 digits
      const hasRightLength = String(phoneNumber).length === 8;

      // saves the phone number if it fulfills all the criteria
      if (isValidNumber && hasRightLength) {
        this._phone = phoneNumber;
        return this._phone;
      } else {
        console.log(invalidPhoneMessage);
      }
    }
```
The original requirements for the homework only mentioned strings and numbers, so all other inputs (including arrays) are rejected outright here.
Thanks again for all your help!!

Reply:
hi @jawkha, great! looks good indeed (only slight issue is indentation :P). 

when you will progress and learn to understand how to write code and solve real world problems, you will also learn more and more to appreciate the value of writing readable and maintainable code. in the end, programming is a group exercise. the easier it is for others (say, your colleagues) it is to reason about things you wrote, the more productive they can be. 

and it's even handy for yourself, because it is very common to work on code you wrote yourself months or years ago.

Reply:
oh yeah, one more thing. 

this:
```js
      if (isValidNumber && hasRightLength) {
        this._phone = phoneNumber;
        return this._phone;
      } else {
        console.log(invalidPhoneMessage);
      }
  }
```

can be written as:
```js
      if (isValidNumber && hasRightLength) {
        this._phone = phoneNumber;
        return this._phone;
      }
      console.log(invalidPhoneMessage);
  }
```

Reply:
Sorry for the wrong indentation. I thought I had taken care of it before making the commit :(
```js
  set phone(phone) {
    const invalidPhoneMessage = "Please input a valid phone number.";
    // only accepts inputs with type of string or number
    if (typeof phone !== "string" || typeof phone !== "number") {
      console.log(invalidPhoneMessage);
      return;
    }
    // changes the type of input to Number
    const phoneNumber = Number(phone);

    // to check that the input after conversion is a valid positive integer
    const isValidNumber =
      typeof phoneNumber === "number" &&
      Number.isInteger(phoneNumber) &&
      phoneNumber > 0;

    // to check that the number has 8 digits
    const hasRightLength = String(phoneNumber).length === 8;

    // saves the phone number if it fulfills all the criteria
    if (isValidNumber && hasRightLength) {
      this._phone = phoneNumber;
      return this._phone;
    }
    console.log(invalidPhoneMessage);
  }
```
Thank you again for your feedback. They have been really helpful. And sorry again for the sloppiness. To not be a drain on my mentors and colleagues' time, I have to work more diligently on writing cleaner code.

Reply:
don't fret it, you're doing a great job!